api/src/main/java/io/onetable/spi/sync/TargetClient.java

```diff
@@ -85,4 +88,10 @@ public interface TargetClient {
 
   /** Returns the {@link io.onetable.model.storage.TableFormat} the client syncs to */
   TableFormat getTableFormat();
+
+  /** Returns the Table Format name as a Sting */
+  String getTableFormatName();
```

Comment:
Is associating the implementation with a specific `TableFormatName` really needed. Consider a future scenario when support for multiple versions of `HUDI` is provided through two implementations of `HUDI`. What will be the table name hard coded in the two implementations. For simplicity both can return `HUDI`. But that has the potential for conflicts in some cases. If they return different names, then a map of all the names has to be maintained.
On the other hand, consider the current design. It allows linking any table format name to a specific implementation. For e.g. here https://github.com/onetable-io/onetable/blob/7d40c65831197412644d4732d6757b5919f150a1/utilities/src/main/resources/onetable-client-defaults.yaml#L14, `HUDI` is the table format identifier given to the default implementation. This config can be overwritten at run time. This appears more flexible to me.

@lmccay, what are your thoughts on using the current approach?

Reply:
Hi @ashvina - Thanks for your insights and pointer. I don't quite follow the statement that a map of all the names needs to be maintained. If you mean as a developer that you need to make sure that your super secret extension is unique, then yes you do. It would be trivial to add a command that lists all the Client format names that are currently in your classpath via ServiceLoader. If what you mean is that there is some external mapping needed to resolve a classname from a format name then, no it is done in the code and you get that mapping for free. The name being codified into the class and the class being discoverable from the classpath as implementing the intended interface is all you need.

I mentioned versioning as something that was bugging me as well (maybe in one of the issues) even with the existing approach (enum). The names can easily be augmented with a version when there is an incompatible change. HUDI, HUDI-2, HUDI-3 where the versions would make sense with the actual Hudi versions maybe.

The use of a config file is indeed another way to do this but configuring classnames is a cumbersome and error prone practice where as the ServiceLoader approach relies on the developer providing the classes name within the jar and service loader config. I've been trying to avoid configuration with classnames for years now. A classname changes in a new release and you have to find it in all config files on all nodes of a cluster or create upgrade procedures or tooling to replace the classnames or config in general. This gets messy.

It looks like the config is identifying the actual targetClients as TargetClientProviders which I didn't think was a thing. Maybe I am missing something.

All of that said, I am not married to this and just found it a fun way to contribute to the extensibility.

Reply:
> The use of a config file is indeed another way to do this but configuring classnames is a cumbersome and error prone practice where as the ServiceLoader approach relies on the developer providing the classes name within the jar and service loader config.
I agree, and I like `ServiceLoader` approach more than the current config file approach in general. However, the config file allows a lot of room for customization; a user can specific the exact provider for an implementation.

I have a suggestion to marry the strengths of both these approaches, let me know what you think. Keep the `getFormatName` method as you have added in the PR. But do not assume it to be unique. Which means, add a mechanism to select one of the two implementations that have provide the same name to `TableFormatClientFactory.createTargetClientForName()`. Say for e.g. Hudi and Hud2 both return the name as HUDI. Then in the dataset config, allow the user to provide the specific implementation class, say to link HUDI to Hudi2.
`for e.g sourceFormat: HUDI   ---> sourceFormat: HUDI:x.y.z.Hudi2`
I am eager to hear your thoughts on improving this. 
  
  
> make sure that your super secret extension is unique

No there is no super-secret extension :-)

Reply:
I don't see more flexibility in either approach to be honest. It seems like you are trying to add in a tie breaker when implementations have the same name accidentally. Doing it on purpose and requiring a config file to override it doesn't seem like something anyone would want to do, assuming they understand the integration/extension model.

So, let's step back, ServiceLoader allows you to load all the implementations of a given interface from the classpath. It doesn't matter where on the classpath they exist. It doesn't require a rebuild in order to add a new jar to the classpath of a given application. Once they are loaded, you have a set of implementation classes that you know the interface for. In usecases like this where you generally want to declaratively specify the implemention to use, it is commonly done by using a method within the interface to uniquely identify each implementation so that it can be requested at runtime.

In this particular PR, it is assumed that the unique implementations are specified at runtime through the perTableConfig by setting the names to use. But in other usecases, such as in Apache Knox, we use the names in deployment descriptors which are similar to a config file approach. We happen to generate JEE applications from it and deploy them.

If we were to have a config file that drove the implementation and versions to use for a given sync operation then you could consume that config to get the names for that sync and interrogate the set of ServiceLoader implementations to find those specific implementations. You could also intercept such a pattern with an override of the classname to use as the implementation and skip the ServiceLoader implementation and create it via reflection. (or be way more complicated about the filtering of ServiceLoader implementations by looking for the classname that matches - I see no value in this).

Now, one thing that is quite interesting about the config file approach that is missing from the ServiceLoader is the config params. This leads me to think more about this as a means to add the required config to the ServiceLoader identified implementations than driving the implementations to use. So, the factory that is using the ServiceLoader could then use some reflection based injection or add it to the perTableConfig or configuration objects that are passed in through init(). I wasn't aware that there were implementation specific config requirements from looking at the existing factory class.

Anyway, those are my thoughts on it. I have a very specific perspective on how to use ServiceLoader based on my experience though and I may be not seeing an additional synergy between the two approaches.

Reply:
> No there is no super-secret extension :-)

The super secret extension that I was referring to would be something like a proprietary format used by an enduser or by tooling or enterprise solutions that may not be added to onetable itself. I see this as the most likely near-term extension usecase.

So, consider a DW vendor that has a onetable extension for their product. They would provide a library download with the core onetable binaries but also with a tooling specific extension. Their proprietary table format is now magically compatible with HUDI, ICEBERG, DELTA.

Reply:
Thank you for the thorough explanation. I appreciate your effort.
I already support this change. Maybe I didn't express myself clearly enough. Let me try again. If it still doesn't work, I would rather let you proceed so we don't hold up the progress.

> it is commonly done by using a method within the interface to uniquely identify each implementation

I have a question about this ^ that you wrote. How can you ensure the uniqueness of the service name? Different companies may develop their own implementations for the `TargetClients` for various table formats, and they may all return HUDI, ICEBERG, or DELTA as the service name. For example, suppose Microsoft and Google both provide their own implementations for HUDI, ICEBERG, and DELTA TargetClients and publish them in a single jar. Suppose company XYZ wants to use the default HUDI source, Microsoft's DELTA source, and Google's ICEBERG source. If XYZ adds all the jars to the classpath, there will be nine implementations with many duplicates. How will the utility handle this situation? Should it use a tie breaker or throw an error? I think using the fully qualified class name would be a better option.

> If we were to have a config file that drove the implementation 

I did not propose to use the config files along with META-INF for the tie breaker. Instead, I wanted to be able to specify the class FQN in the `perTableConfig` in case of name conflicts.

Reply:
>  For example, suppose Microsoft and Google both provide their own implementations for HUDI, ICEBERG, and DELTA TargetClients and publish them in a single jar.

@ashvina - I understand. I may just be missing why msft and goog would do that. If the extensibility guidelines explain that you need to provide a unique name and there are already builtin implementations.

If there are some company specific extensions to the open formats, that seems contrary to the whole thing but I also think it would be obvious to call them MSFT-HUDI and GOOG-HUDI instead of HUDI.

Anyway, having a separate configured list of FQCN's to address this and accommodate it within the filtering of the ServiceLoader filtering wouldn't be difficult to do.

Reply:
> I may just be missing why msft and goog would do that.

Independent publishers may not do this intentionally. It can occur by accident. This issue is not very serious at the moment.



> accommodate it within the filtering of the ServiceLoader filtering wouldn't be difficult to do.
I agree.